runtime/parachains/src/scheduler.rs

```diff
@@ -13,3 +13,1403 @@
+					while prev_scheduled_in_order.peek().map_or(
+						false,
+						|(_, assign)| (assign.core.0 as usize) < core_index,
+					) {
+						let _ = prev_scheduled_in_order.next();
+					}
+
+					// check the first entry already scheduled with core index >= than the one we
+					// are looking at. 3 cases:
+					//  1. No such entry, clearly this core is not scheduled, so we need to schedule and put at the end.
+					//  2. Entry exists and has same index as the core we are inspecting. do not schedule again.
+					//  3. Entry exists and has higher index than the core we are inspecting. schedule and note
+					//     insertion position.
+					prev_scheduled_in_order.peek().map_or(
+						Some(scheduled.len()),
+						|(idx_in_scheduled, assign)| if (assign.core.0 as usize) == core_index {
+							None
+						} else {
+							Some(*idx_in_scheduled)
+						},
+					)
+				};
+
+				let schedule_and_insert_at = match schedule_and_insert_at {
+					None => continue,
+					Some(at) => at,
+				};
+
+				let core = CoreIndex(core_index as u32);
+
+				let core_assignment = if core_index < parachains.len() {
+					// parachain core.
+					Some(CoreAssignment {
+						kind: AssignmentKind::Parachain,
+						para_id: parachains[core_index],
+						core: core.clone(),
+						group_idx: Self::group_assigned_to_core(core, now)
+							.expect("core is not out of bounds and we are guaranteed \
+									to be after the most recent session start; qed"),
+					})
+				} else {
+					// parathread core offset, rel. to beginning.
+					let core_offset = (core_index - parachains.len()) as u32;
+
+					parathread_queue.take_next_on_core(core_offset).map(|entry| CoreAssignment {
+						kind: AssignmentKind::Parathread(entry.claim.1, entry.retries),
+						para_id: entry.claim.0,
+						core: core.clone(),
+						group_idx: Self::group_assigned_to_core(core, now)
+							.expect("core is not out of bounds and we are guaranteed \
+									to be after the most recent session start; qed"),
+					})
+				};
+
+				if let Some(assignment) = core_assignment {
+					scheduled_updates.push((schedule_and_insert_at, assignment))
+				}
+			}
+
+			// at this point, because `Scheduled` is guaranteed to be sorted and we navigated unassigned
+			// core indices in ascending order, we can enact the updates prepared by the previous actions.
+			//
+			// while inserting, we have to account for the amount of insertions already done.
+			//
+			// This is O(n) as well, capped at n operations, where n is the number of cores.
+			for (num_insertions_before, (insert_at, to_insert)) in scheduled_updates.into_iter().enumerate() {
+				let insert_at = num_insertions_before + insert_at;
+				scheduled.insert(insert_at, to_insert);
+			}
+
+			// scheduled is guaranteed to be sorted after this point because it was sorted before, and we
+			// applied sorted updates at their correct positions, accounting for the offsets of previous
+			// insertions.
+		}
+
+		Scheduled::set(scheduled);
```

Comment:
if the validator group is empty, should this not be set to an empty vec too?

Reply:
If there are no validator groups, we return `None` earlier. Actually I think this handles the case of some validator groups being empty somewhat gracefully, although parathreads will count as having been retried even when there was nobody to back the candidate. I can amend that if desired